sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -47,6 +47,45 @@ case class Md5(child: Expression) extends UnaryExpression with ExpectsInputTypes
 }
 
 /**
+ * A function that returns a hash value of the argument
+ */
+case class Hash(children: Expression*) extends Expression {
+
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = children(0).nullable
+  override def dataType: DataType = IntegerType
+
+  override def eval(input: InternalRow): Any = {
+    val arglist = children.map(_.eval(input).asInstanceOf[AnyRef])
+    hashCode(arglist)
+  }
+
+  def hashCode(v: Any): Int = v match {
+    case null => 0
+    case arr: Array[_] => {
+      arr.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case seq: Seq[_] => {
+      seq.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case m: Map[_, _] => {
+      var res: Int = 0
+      m.foreach { case (k, v) => res += (hashCode(k) + hashCode(v)) }
```

Comment:
The ordering of the map keys in `foreach` is not stable, so I think this may lead to unstable hashcodes for identical maps.